### ISSUE TITLE
False positives for missing CONFIG_MITIGATION_ on older kernels

### ISSUE THREAD

**Issue by asarubbo** [a13xp0p0v/kernel-hardening-checker#219]
Hello,

While running v0.6.17.1 on a 6.1.170 kernel configuration, I observed the following:

```
CONFIG_MITIGATION_SPECTRE_V1          |kconfig| self_protection  |defconfig |     y      | FAIL: is not found
CONFIG_MITIGATION_SPECTRE_V2          |kconfig| self_protection  |defconfig |     y      | FAIL: is not found
```

However, these options are not present in kernel 6.1.x and were introduced later (e.g., they appear in 7.0.3).

Is there any mechanism in place to filter out configuration checks based on the target kernel version, so that options not available in a given version are not reported as missing?

**Comment by a13xp0p0v** [comment]
Hello @asarubbo.

Thanks for creating this issue.

The `kernel-hardening-checker` tool recommends the latest security hardening options.

Supporting a kernel version range for each of the Kconfig / cmdline /sysctl parameter in `kernel-hardening-checker` would be extremely complicated.

Moreover, different GNU/Linux distributions have different kernel versioning policy that may not match to the vanilla kernel.

However, I understand your request to mute some checks.

Please see the the issue https://github.com/a13xp0p0v/kernel-hardening-checker/issues/50, where we discussed various approaches to solve this. 

What do you think?

**Comment by asarubbo** [comment]
I haven’t analyzed all possible scenarios, but I think there are two categories of users:

- those who compile a kernel on a constrained distro (e.g. Gentoo - excluding the recent availability of binary kernels)
- those who compile a kernel on a stable/release distro (like Rocky) to stay up to date (well, I did, because they slept on fixing the copyfail issue)

In both cases, I believe the simplest approach would be to rely on the upstream kernel. This also relates to the issue that we know when a `CONFIG_FOO` was introduced upstream, but we don’t necessarily know if and when it was introduced downstream.

I haven’t looked at the code, and I’m not a Python developer anyway, so with that premise in mind, what do you think about adding a column that statically records the first upstream kernel version where a given config option appeared? Without any additional runtime or command-line checks.

In this way, when checking a config on e.g. 6.18, if `kernel-hardening-checker` reports that `CONFIG_FOO` exists in 6.18 but I cannot find it in my kernel config, I can be confident that something is wrong on my side, rather than concluding that “this config does not exist in that Linux release”.

What do you think?

**Comment by a13xp0p0v** [comment]
> what do you think about adding a column that statically records the first upstream kernel version where a given config option appeared? Without any additional runtime or command-line checks.

Oh, there are wonderful projects that solve this task much better comparing to an additional column in the `kernel-hardening-checker` output!

Let me quote the Q&A in the README: https://github.com/a13xp0p0v/kernel-hardening-checker#questions-and-answers

**Q: Can I easily check which kernel versions support some Kconfig option?**

**A: Yes. See the [LKDDb](https://cateee.net/lkddb/web-lkddb/) project (Linux Kernel Driver Database) by Giacomo Catenazzi [@cateee](https://github.com/cateee/lkddb). You can use it for the mainline or stable tree from [kernel.org](https://kernel.org/) or for your custom kernel sources.**

It's a nice project, you gonna like it.

**Comment by a13xp0p0v** [comment]
Closing for now. Hope to implement check filtering in #50. Looking forward to the contributions.
